smithy-model/src/main/java/software/amazon/smithy/model/knowledge/NullableIndex.java

```diff
@@ -17,21 +17,55 @@
 
 import java.lang.ref.WeakReference;
 import java.util.Objects;
+import java.util.Set;
 import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.loader.ModelAssembler;
 import software.amazon.smithy.model.shapes.MemberShape;
 import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.shapes.ShapeType;
 import software.amazon.smithy.model.shapes.ToShapeId;
+import software.amazon.smithy.model.traits.BoxTrait;
 import software.amazon.smithy.model.traits.ClientOptionalTrait;
 import software.amazon.smithy.model.traits.DefaultTrait;
 import software.amazon.smithy.model.traits.InputTrait;
 import software.amazon.smithy.model.traits.RequiredTrait;
 import software.amazon.smithy.model.traits.SparseTrait;
+import software.amazon.smithy.utils.SetUtils;
 
 /**
  * An index that checks if a member is nullable.
+ *
+ * <p>Note: this index assumes Smithy 2.0 nullability semantics.
+ * There is basic support for detecting 1.0 models by detecting
+ * when a removed primitive prelude shape is targeted by a member.
+ * Beyond that, 1.0 models SHOULD be loaded through a {@link ModelAssembler}
+ * to upgrade them to IDL 2.0.
  */
 public class NullableIndex implements KnowledgeIndex {
 
+    private static final Set<ShapeId> V1_REMOVED_PRIMITIVE_SHAPES = SetUtils.of(
+            ShapeId.from("smithy.api#PrimitiveBoolean"),
+            ShapeId.from("smithy.api#PrimitiveByte"),
+            ShapeId.from("smithy.api#PrimitiveShort"),
+            ShapeId.from("smithy.api#PrimitiveInteger"),
+            ShapeId.from("smithy.api#PrimitiveLong"),
+            ShapeId.from("smithy.api#PrimitiveFloat"),
+            ShapeId.from("smithy.api#PrimitiveDouble"));
+
+    private static final Set<ShapeType> V1_INHERENTLY_BOXED = SetUtils.of(
+            ShapeType.STRING,
+            ShapeType.BLOB,
+            ShapeType.TIMESTAMP,
+            ShapeType.BIG_DECIMAL,
+            ShapeType.BIG_INTEGER,
+            ShapeType.LIST,
+            ShapeType.SET,
```

Comment:
Was surprised to see this without a suppression for `ShapeType.SET` being deprecated, is that missing?

Reply:
I decided not to deprecate this enum members. We can deprecate the SetShape and pretend it doesn't exist because it extends from ListShape. You can't do that with enum members in Java, so you still need to write type handling code that accounts for SET. I didn't see any value in deprecating it.